aten/src/ATen/native/transformers/attention.cpp

```diff
@@ -710,6 +710,28 @@ std::tuple<Tensor, Tensor> _scaled_dot_product_attention(
       query_, key, value, attn_mask_, dropout_p, is_causal);
 }
 
+inline void validate_sdpa_input(
+    const Tensor& query_,
+    const Tensor& key,
+    const Tensor& value,
+    const c10::optional<Tensor>& attn_mask_,
+    double dropout_p,
+    bool is_causal) {
+  TORCH_CHECK(
+      query_.dtype() == key.dtype() && query_.dtype() == value.dtype(),
+      "Query, key, and value must have the same dtype");
+  TORCH_CHECK(
+      query_.device() == key.device() && query_.device() == value.device(),
+      "Query, key, and value must be on the same device");
+  TORCH_CHECK(
+      query_.dim() >= 2 && key.dim() >= 2 && value.dim() >= 2,
+      "Query, key, and value must have at least 2 dimensions");
+  if (attn_mask_.has_value()){
+    auto mask_dtype = attn_mask_->dtype();
+    TORCH_CHECK(mask_dtype == at::kBool || mask_dtype == query_.dtype(),"attn_mask dtype must either match query dtype or be dtype bool.")
```

Comment:
nit: In general it's a better user experience if you say what you got instead.

So you could say "Expected attn_mask dtype to match query dtype, but got ", attn_mask.dtype(), " and ", query.dtype(), " instead."

I know this is an annoying nit and it also applies to all your other checks, but it can make it easier for a user to spot the error in their code without having to drop into a debugger.

Reply:
Good point, I updated